stdlib/public/core/StringComparison.swift

```diff
@@ -0,0 +1,1368 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftShims
+
+ @inline(__always)
+ @_inlineable
+ public func _explodeBits(_ core: _StringGuts) -> _StringGuts._RawBitPattern {
+   return (
+     Builtin.reinterpretCast(core._object),
+     Builtin.reinterpretCast(core._otherBits))
+ }
+ @inline(__always)
+ @_inlineable
+ public
+ func _implodeBits(_ bits: _StringGuts._RawBitPattern) -> _StringGuts {
+   return _StringGuts(
+     object: Builtin.reinterpretCast(bits.0),
+     otherBits: Builtin.reinterpretCast(bits.1))
+  }
+
+@inline(never)
+@effects(readonly)
+public
+func _compareHack(
```

Comment:
What is the hack? Why is it there? Please provide some annotation or otherwise. Also, comparison was already hacking around ARC, so how is this approach different?